apollo-entity/src/main/java/com/spotify/apollo/entity/CodecEntityMiddleware.java

```diff
@@ -0,0 +1,250 @@
+  }
+
+  @Override
+  public <R> Middleware<AsyncHandler<R>, AsyncHandler<Response<ByteString>>>
+  asyncSerializerDirect(Class<? extends R> entityClass) {
+    return inner -> inner
+        .map(this::ensureResponse)
+        .map(serialize(entityClass));
+  }
+
+  @Override
+  public <R> Middleware<AsyncHandler<Response<R>>, AsyncHandler<Response<ByteString>>>
+  asyncSerializerResponse(Class<? extends R> entityClass) {
+    return inner -> inner.map(serialize(entityClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityHandler<E, E>, SyncHandler<Response<ByteString>>>
+  direct(Class<? extends E> entityClass) {
+    return direct(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityHandler<E, R>, SyncHandler<Response<ByteString>>>
+  direct(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        deserialize(entityClass)
+            .flatMap(r -> ctx -> mapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityResponseHandler<E, E>, SyncHandler<Response<ByteString>>>
+  response(Class<? extends E> entityClass) {
+    return response(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityResponseHandler<E, R>, SyncHandler<Response<ByteString>>>
+  response(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        deserialize(entityClass)
+            .flatMap(r -> ctx -> flatMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityAsyncHandler<E, E>, AsyncHandler<Response<ByteString>>>
+  asyncDirect(Class<? extends E> entityClass) {
+    return asyncDirect(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityAsyncHandler<E, R>, AsyncHandler<Response<ByteString>>>
+  asyncDirect(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        Middleware.syncToAsync(deserialize(entityClass))
+            .flatMap(r -> ctx -> asyncMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  @Override
+  public <E> Middleware<EntityAsyncResponseHandler<E, E>, AsyncHandler<Response<ByteString>>>
+  asyncResponse(Class<? extends E> entityClass) {
+    return asyncResponse(entityClass, entityClass);
+  }
+
+  @Override
+  public <E, R> Middleware<EntityAsyncResponseHandler<E, R>, AsyncHandler<Response<ByteString>>>
+  asyncResponse(Class<? extends E> entityClass, Class<? extends R> entityResponseClass) {
+    return inner ->
+        Middleware.syncToAsync(deserialize(entityClass))
+            .flatMap(r -> ctx -> asyncFlatMapPayload(r, inner.invoke(ctx)))
+            .map(serialize(entityResponseClass));
+  }
+
+  private <E> SyncHandler<Response<E>> deserialize(Class<? extends E> entityClass) {
```

Comment:
But in the case that things go right, the payload of the 'response' created during deserialization is in fact the *request* payload, right? That's the thing that looks odd. Calling something a response when it's in fact the request. I was totally confused when reading the deserialize method, because it seemed to me that the codec would have to apply the endpoint handler function and I didn't see how that could work.

I'm more concerned about future maintenance than users. From an API perspective, I really like this PR. It's just that it's hard (for me) to navigate this code.